extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -142,16 +147,27 @@ export class Bind {
    * @private
    */
   initialize_() {
-    this.scanPromise_ = this.scanBody_(this.ampdoc.getBody());
-    this.scanPromise_.then(results => {
-      const {boundElements, bindings, expressionToElements} = results;
+    this.addBindingsForSubtree(this.ampdoc.getBody());
+  }
 
-      this.boundElements_ = boundElements;
+  /**
+   * Scans the substree rooted at `rootElement` and adds bindings for nodes
+   * that contain bindable elements. This function is not idempotent. To remove
+   * bindings for a subtree, see #removeBindingsForSubtree. Returns a promise
+   * that resolves after bindings have been added.
+   * @param {!Element} rootElement
+   * @return {Promise}
+   */
+  addBindingsForSubtree(rootElement) {
+    this.scanPromise_ = this.scanSubtree_(rootElement).then(results => {
+      const {boundElements, bindings, expressionToElements} = results;
 
+      this.boundElements_ = this.boundElements_.concat(boundElements);
       Object.assign(this.expressionToElements_, expressionToElements);
+      this.bindings_ = this.bindings_.concat(bindings);
 
-      this.evaluator_ = new BindEvaluator();
-      const parseErrors = this.evaluator_.setBindings(bindings);
+      this.evaluator_ = this.evaluator_ || new BindEvaluator();
+      const parseErrors = this.evaluator_.setBindings(this.bindings_);
```

Comment:
This will cause the evaluator to re-parse every expression, and expression parsing currently is the slowest operation in amp-bind. Instead, we should support incrementally add of expressions.

You can probably avoid the `this.bindings_` variable, which saves work when removing bindings from a subtree.